### PR TITLE
removed ADD SUBTRACT and MULTIPLY

### PR DESCRIPTION
**Pull request by winitzki**
Simplify code. There are no more `Add`, `Subtract`, and `Multiply` case classes; `Sum` and `Product` are used instead. Functionality is unchanged.

### CODE REVIEW

File: src/main/scala/com/github/symcal/Expr.scala [hunk 1]
```diff
@@ -261,28 +267,38 @@ final case class Sum(es: Expr*) extends Expr {
...
-    val mergedConstants = const.foldLeft[Expr](Const(0))((x, y) ⇒ Const(x.toInt + y.toInt))
-    val mergedExprs = mergedConstants match {
-      case Const(0) ⇒ nonconst
-      case _ ⇒ nonconst :+ mergedConstants
+    val (constants, nonconstants) = es.map(_.simplify).partition(_.isConst)
+    val nonconstantsFlattened: Seq[Expr] = nonconstants.flatMap {
+      case Sum(es@_*) ⇒ es
+      case x => Seq(x) // not a sum
+    }
+    // mergedConstants is the product of all constants in the list; also could be 0 or 1.
```
Comment (hamdi-jenzri) [inline src/main/scala/com/github/symcal/Expr.scala]: `mergedConstants is the sum of all constants`?

File: src/main/scala/com/github/symcal/Expr.scala [hunk 2]
```diff
@@ -261,28 +267,38 @@ final case class Sum(es: Expr*) extends Expr {
...
+    val mergedExprs: Seq[Expr] = mergedConstants match {
+      case 0 ⇒ nonconstantsFlattened
+      case _ ⇒ nonconstantsFlattened :+ Const(mergedConstants)
     }
     // There are three cases now: empty sequence, one expr, and more than one expr.
     mergedExprs.headOption match {
       case Some(e) ⇒
         if (mergedExprs.length == 1) {
-          // In this case, the simplified result is not a `Sum`.
+          // In this case, the simplified result is not a `Product`.
```
Comment (hamdi-jenzri) [inline src/main/scala/com/github/symcal/Expr.scala]: what do you mean by `// In this case, the simplified result is not a `Product`.`?

File: src/main/scala/com/github/symcal/Expr.scala [hunk 3]
```diff
@@ -298,23 +314,30 @@ final case class Product(es: Expr*) extends Expr {
...
-      }
+    val (constants, nonconstants) = es.map(_.simplify).partition(_.isConst)
+    val nonconstantsFlattened: Seq[Expr] = nonconstants.flatMap {
+      case Product(es@_*) ⇒ es
+      case x => Seq(x) // not a product
+    }
+    // mergedConstants is the product of all constants in the list; also could be 0 or 1.
+    val mergedConstants: Int = constants
+      .collect { case x@Const(_) ⇒ x } // This converts into Seq[Const]. We know that we are not losing any values here.
+      .map(_.value) // Here we can just get _.value of the `Const`.
```
Comment (hamdi-jenzri) [inline src/main/scala/com/github/symcal/Expr.scala]: couldn't we have done `x.value` in the collect?

File: src/main/scala/com/github/symcal/Expr.scala (hunk 2 above)
Comment (winitzki) [inline src/main/scala/com/github/symcal/Expr.scala]: I will clarify this.

File: src/main/scala/com/github/symcal/Expr.scala (hunk 3 above)
Comment (winitzki) [inline src/main/scala/com/github/symcal/Expr.scala]: Yes, you are right.

File: src/main/scala/com/github/symcal/Expr.scala (hunk 1 above)
Comment (winitzki) [inline src/main/scala/com/github/symcal/Expr.scala]: I will fix the comments, they are mismatched after copy/paste.

**Pull request merged**